native/cocos/renderer/gfx-vulkan/VKBuffer.cpp

```diff
@@ -41,87 +41,80 @@ CCVKBuffer::~CCVKBuffer() {
-    _gpuBuffer->memUsage = _memUsage;
-    _gpuBuffer->size = _size;
-    _gpuBuffer->stride = _stride;
-    _gpuBuffer->count = _count;
-
-    if (hasFlag(_usage, BufferUsageBit::INDIRECT)) {
-        const size_t drawInfoCount = _size / sizeof(DrawInfo);
-        _gpuBuffer->indexedIndirectCmds.resize(drawInfoCount);
-        _gpuBuffer->indirectCmds.resize(drawInfoCount);
-    }
-
-    cmdFuncCCVKCreateBuffer(CCVKDevice::getInstance(), _gpuBuffer);
-    CCVKDevice::getInstance()->getMemoryStatus().bufferSize += _size;
-    CC_PROFILE_MEMORY_INC(Buffer, _size);
+    createBuffer(_size, _count);
 
-    _gpuBufferView = ccnew CCVKGPUBufferView;
-    createBufferView();
+    createBufferView(_size);
 }
 
 void CCVKBuffer::doInit(const BufferViewInfo &info) {
     auto *buffer = static_cast<CCVKBuffer *>(info.buffer);
     _gpuBuffer = buffer->gpuBuffer();
-    _gpuBufferView = ccnew CCVKGPUBufferView;
-    createBufferView();
+
+    createBufferView(_size);
 }
 
-void CCVKBuffer::createBufferView() {
+void CCVKBuffer::createBuffer(uint32_t size, uint32_t count) {
+    _gpuBuffer = ccnew CCVKGPUBuffer;
+    _gpuBuffer->size = size;
+    _gpuBuffer->count = count;
+
+    _gpuBuffer->usage = _usage;
+    _gpuBuffer->memUsage = _memUsage;
+    _gpuBuffer->stride = _stride;
+    _gpuBuffer->init();
+}
+
+void CCVKBuffer::createBufferView(uint32_t range) {
+    _gpuBufferView = ccnew CCVKGPUBufferView;
+    _gpuBufferView->range = range;
+
     _gpuBufferView->gpuBuffer = _gpuBuffer;
     _gpuBufferView->offset = _offset;
-    _gpuBufferView->range = _size;
-    if (_gpuBufferView->gpuBuffer->vkBuffer) {
-        CCVKDevice::getInstance()->gpuDescriptorHub()->update(_gpuBufferView);
-    }
 }
 
 void CCVKBuffer::doDestroy() {
-    if (_gpuBufferView) {
-        CCVKDevice::getInstance()->gpuDescriptorHub()->disengage(_gpuBufferView);
-        delete _gpuBufferView;
-        _gpuBufferView = nullptr;
-    }
-
-    if (_gpuBuffer) {
-        if (!_isBufferView) {
-            CCVKDevice::getInstance()->gpuBufferHub()->erase(_gpuBuffer);
-            CCVKDevice::getInstance()->gpuRecycleBin()->collect(_gpuBuffer);
-            CCVKDevice::getInstance()->gpuBarrierManager()->cancel(_gpuBuffer);
-            delete _gpuBuffer;
-            CCVKDevice::getInstance()->getMemoryStatus().bufferSize -= _size;
-            CC_PROFILE_MEMORY_DEC(Buffer, _size);
-        }
-        _gpuBuffer = nullptr;
-    }
+    _gpuBufferView = nullptr;
+    _gpuBuffer = nullptr;
 }
 
 void CCVKBuffer::doResize(uint32_t size, uint32_t count) {
-    CCVKDevice::getInstance()->getMemoryStatus().bufferSize -= _size;
-    CC_PROFILE_MEMORY_DEC(Buffer, _size);
-    CCVKDevice::getInstance()->gpuRecycleBin()->collect(_gpuBuffer);
+    createBuffer(size, count);
 
-    _gpuBuffer->size = size;
-    _gpuBuffer->count = count;
-    cmdFuncCCVKCreateBuffer(CCVKDevice::getInstance(), _gpuBuffer);
+    IntrusivePtr<CCVKGPUBufferView> oldBufferView = _gpuBufferView;
```

Comment:
Is it need to create a temp IntrusivePtr object?

Reply:
It doesn't seem necessary. `InputAssembler` or `DescriptrSet` also hold old bufferview.

Reply:
I did some tests. Old bufferview has to be alive during `Hub::update`. Bufferview destroyed during viewSet traversal  may cause iterator invalidation.

Reply:
> Bufferview destroyed during viewSet traversal may cause iterator invalidation.

Where does it happen in update function?

Reply:
It traverse all related InputAssemblers to find the old bufferview, and replace it with the newly created one. 
```cpp
auto iter = _ias.find(oldBuffer);
if (iter != _ias.end()) {
    for (const auto &ia : iter->second) {    //  <---- Traverse iaSet.
        ia->update(oldBuffer, newBuffer);    //  <---- Replace with new bufferview will release reference to the old bufferview. 
                                             //  BufferView removes relations in iaSet after being destroyed and iterator invalidation happens.
        _ias[newBuffer].insert(ia);
    }
    _ias.erase(iter);
}
```

Reply:
The code here is 
```c++
IntrusivePtr<CCVKGPUBufferView> oldBufferView = _gpuBufferView;
    createBufferView(size);
    CCVKDevice::getInstance()->gpuDescriptorHub()->update(oldBufferView, _gpuBufferView);
    CCVKDevice::getInstance()->gpuIAHub()->update(oldBufferView, _gpuBufferView);
```

Which code will do the operations as you mentioned above?

Reply:
```cpp

void CCVKBuffer::doResize(uint32_t size, uint32_t count) {
    // ....
    CCVKDevice::getInstance()->gpuDescriptorHub()->update(oldBufferView, _gpuBufferView);   // <- Traverse viewSet，and invoke set::update(old, new)
}

void CCVKGPUInputAssembler::update(const CCVKGPUBufferView *oldBuffer, const CCVKGPUBufferView *newBuffer) {
    for (uint32_t i = 0; i < gpuVertexBuffers.size(); ++i) {
        if (gpuVertexBuffers[i].get() == oldBuffer) {
            gpuVertexBuffers[i] = newBuffer;      // <- replace with new bufferview, and the old bufferview may be released.
        }
    }
}

void CCVKGPUBufferView::shutdown() {
    CCVKDevice::getInstance()->gpuDescriptorHub()->disengage(this);
    CCVKDevice::getInstance()->gpuIAHub()->disengage(this);     // <- bufferview releases relations in hub, within the traverse.
}
```
DescriptorSetHub will be the same.

Reply:
Got it, then it is better do add some comments, or it may be changed to raw pointer in future.

Reply:
Ok, I am adding comments.